### ISSUE TITLE
Custom Authentication Handler - Unable to retrieve Body

### ISSUE THREAD

**Issue by Zavix210** [ThreeMammals/Ocelot#1233]
## Expected Behavior 
Be able to read the Request.Body from the HTTP Request.

## Actual Behavior / Motivation for New Feature
Unable to retrieve the body. The reason for doing this is I'm implementing an API Key setup with HMAC and would like to verify the body data against the HMAC headers.

##Notes 
I'm unsure if this is by design or not however i can't find a way to actually retrieve the body request from my custom authentication handler. By changing ocelots request body manipulation so that multiple reads can be made by doing:  request.EnableBuffering(). Fix's the problem, unsure if this is not done on purpose due to memory.

**Comment by absentik** [comment]
I have similar problem. I made custom authentication handler (based on `Microsoft.AspNetCore.Authentication.AuthenticationHandler`) which checks api key. 

When I call **controller method** without api key or with wrong api key I got 401 with body.
When I call **ocelot method** without api key or with wrong api key I got 401 without body.
I noticed that in case of ocelot method `HandleChallengeAsync` is not called after `HandleAuthenticateAsync`. 
May be I doing something wrong but I can't find right way how check api key and return message only for the secured routes.

Perhaps it is due with this issue https://github.com/ThreeMammals/Ocelot/issues/1144
